### PR TITLE
Feature/416811 bulk upload errors processing

### PR DESCRIPTION
**Pull request by abidkayanidefra**
Feature/416811 bulk upload errors processing

### CODE REVIEW

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs [hunk 1]
```diff
@@ -1,24 +1,37 @@
...
 
-        // All subsidiaries with an org id, where no relationship already exists
-        var knownSubsidiariesToAdd = subsidiariesAndOrg.Where(co => co.SubsidiaryOrg != null)
+        // before going further. here we can check with RPD if the company name in the file is matching the RPD Org table company name
+        // TO DO Use this subsidiaries that already have orgs
+        var subsidiariesAndOrgWithValidName = subsidiariesAndOrg
+            .Where(sub => sub.Subsidiary.companies_house_number == sub.SubsidiaryOrg.companiesHouseNumber && sub.Subsidiary.organisation_name == sub.SubsidiaryOrg.name);
+
+        // after checking the name only that collection will pass on for relationship check
+
```
Comment (BrianKnightEviden) [inline src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs:32]: remove extra line

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs [hunk 2]
```diff
@@ -1,24 +1,36 @@
...
+    private const string SubsidiaryBulkUploadMismatchedErrors = "Subsidiary bulk upload errors. Mismatched subsidiaries.";
     private readonly ILogger<BulkSubsidiaryProcessor> _logger = logger;
     private readonly ISubsidiaryService organisationService = organisationService;
     private readonly ICompaniesHouseDataProvider companiesHouseDataProvider = companiesHouseDataProvider;
+    private readonly INotificationService _notificationService = notificationService;
 
-    public async Task Process(IEnumerable<CompaniesHouseCompany> subsidiaries, CompaniesHouseCompany parent, OrganisationResponseModel parentOrg, Guid userId)
+    public async Task Process(IEnumerable<CompaniesHouseCompany> subsidiaries, CompaniesHouseCompany parent, OrganisationResponseModel parentOrg, UserRequestModel userRequestModel)
     {
+        // these are the subsidiaries which exists in RPD
```
Comment (RamakrishnaGujjula) [inline src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs:21]: Remove the comments?

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs [hunk 3]
```diff
@@ -1,24 +1,36 @@
...
             .ToAsyncEnumerable()
             .SelectAwait(async subsidiary => (Subsidiary: subsidiary, SubsidiaryOrg: await organisationService.GetCompanyByCompaniesHouseNumber(subsidiary.companies_house_number)));
 
-        // All subsidiaries with an org id, where no relationship already exists
-        var knownSubsidiariesToAdd = subsidiariesAndOrg.Where(co => co.SubsidiaryOrg != null)
+        // before going further. here we can check with RPD if the company name in the file is matching the RPD Org table company name
+        var subsidiariesAndOrgWithValidName = subsidiariesAndOrg
+            .Where(sub => sub.Subsidiary.companies_house_number == sub.SubsidiaryOrg.companiesHouseNumber && sub.Subsidiary.organisation_name == sub.SubsidiaryOrg.name);
+
+        // after checking the name only that collection will pass on for relationship check
```
Comment (RamakrishnaGujjula) [inline src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs:30]: Comments need to be removed!

File: src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs [hunk 4]
```diff
@@ -147,4 +144,29 @@ public async Task<bool> GetSubsidiaryRelationshipAsync(int parentOrganisationId,
...
+
+        var response = await _httpClient.PostAsync(OrganisationAddSubsidiaryUri, httpContent);
+
+        if (!response.IsSuccessStatusCode)
+        {
+            var problemDetails = await response.Content.ReadFromJsonAsync<ProblemDetails>();
+
+            if (problemDetails != null)
+            {
+                // _logger.LogError("Failed to add subsidiary relationship for Parent: {Parent} Subsidiary: {Subsidiary}", subsidiaryAddModel.ParentOrganisationId, subsidiaryAddModel.ChildOrganisationId);
```
Comment (RamakrishnaGujjula) [inline src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs:161]: Remove the logger if no longer needed.

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs (hunk 2 above)
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs:21]: there are lots of comments between line 20 and 82. its quite complex code. these comments would be useful for anyone reading the code or taking over this application.

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs [hunk 5]
```diff
@@ -28,20 +41,97 @@ public async Task Process(IEnumerable<CompaniesHouseCompany> subsidiaries, Compa
...
+         */
+        var subsidiariesAndOrgWith_InValidName = subsidiariesAndOrg
+            .Where(sub => sub.Subsidiary.companies_house_number == sub.SubsidiaryOrg.companiesHouseNumber && sub.Subsidiary.organisation_name != sub.SubsidiaryOrg.name);
+
+        // check and report subsidiaries with mismatched names.
+        var subWithInvalidName = await subsidiariesAndOrgWith_InValidName.Select(s => s.Subsidiary).ToListAsync();
+        await ReportCompanies(subWithInvalidName, userRequestModel);
+
         // Subsidiaries which do not exist in the RPD
+        // TO DO use these new subsidiaries.
```
[the remark this answers is not in the document]
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs]: Done

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs:31]: done

File: src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs [hunk 6]
```diff
@@ -112,14 +113,10 @@ public async Task<bool> GetSubsidiaryRelationshipAsync(int parentOrganisationId,
...
             }
         }
 
-        response.EnsureSuccessStatusCode();
-        var result = await response.Content.ReadAsStringAsync();
-        return result;
+        return response.StatusCode;
     }
 
     public async Task<string?> AddSubsidiaryRelationshipAsync(SubsidiaryAddModel subsidiaryAddModel)
```
[the remark this answers is not in the document]
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs:122]: done

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs (hunk 3 above)
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs:30]: sorry these comments are required for the logic understanding

File: src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs (hunk 4 above)
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs:161]: code updated.

File: src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs
[the remark this answers is not in the document]
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs:1]: removed exceptions

File: src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs [hunk 7]
```diff
@@ -147,4 +144,29 @@ public async Task<bool> GetSubsidiaryRelationshipAsync(int parentOrganisationId,
 
         return result;
     }
+
+    public async Task<List<OrganisationResponseModel>> GetNoneProccessedCompanies(IEnumerable<CompaniesHouseCompany> subsidiaries)
+    {
+        string json = JsonConvert.SerializeObject(subsidiaries);
```
[the remark this answers is not in the document]
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs:150]: code updated

File: src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs [hunk 8]
```diff
@@ -147,4 +144,29 @@ public async Task<bool> GetSubsidiaryRelationshipAsync(int parentOrganisationId,
...
+            var problemDetails = await response.Content.ReadFromJsonAsync<ProblemDetails>();
+
+            if (problemDetails != null)
+            {
+                // _logger.LogError("Failed to add subsidiary relationship for Parent: {Parent} Subsidiary: {Subsidiary}", subsidiaryAddModel.ParentOrganisationId, subsidiaryAddModel.ChildOrganisationId);
+                throw new ProblemResponseException(problemDetails, response.StatusCode);
+            }
+        }
+
+        response.EnsureSuccessStatusCode();
```
[the remark this answers is not in the document]
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs]: done

File: src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs [hunk 9]
```diff
@@ -87,8 +88,10 @@ public async Task<bool> GetSubsidiaryRelationshipAsync(int parentOrganisationId,
 
             if (problemDetails != null)
             {
-                throw new ProblemResponseException(problemDetails, response.StatusCode);
+                _logger.LogError("Error occurred in GetCompanyByCompaniesHouseNumber call: {Parent} Subsidiary: {Subsidiary}", problemDetails.Detail, response.StatusCode);
```
Comment (mikeymonster) [inline src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs:91]: Check what you are logging in all of these methods - this code puts the problem detail into the {Parent} placeholder and the status code as {Subsidiary}.

**Review by RamakrishnaGujjula — APPROVED** [review]

File: src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs
Comment (BrianKnightEviden) [inline src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs:1]: GetSubsidiaryRelationshipAsync still throws exceptions

**Review by deorab — APPROVED** [review]

File: src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs (hunk 9 above)
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/SubsidiaryService.cs]: fixed.

**Review by deorab — DISMISSED** [review]

**Review by deorab — APPROVED** [review]

**Review by RamakrishnaGujjula — APPROVED** [review]

**Review by deorab — APPROVED** [review]

**Review by mikeymonster — APPROVED** [review]

File: src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs [hunk 10]
```diff
@@ -28,20 +41,97 @@ public async Task Process(IEnumerable<CompaniesHouseCompany> subsidiaries, Compa
...
+            };
+
+            notificationErrorList.Add(newError);
+        }
+
+        var keyErrors = userRequestModel.GenerateKey(SubsidiaryBulkUploadMismatchedErrors);
+        var key = userRequestModel.GenerateKey(SubsidiaryBulkUploadMismatchedProgress);
+        _notificationService.SetStatus(key, "Started reporting invalid subsidiaries.");
+        _notificationService.SetErrorStatus(keyErrors, notificationErrorList);
+        _logger.LogInformation("Mismatched named subsidiaries found. Subsidiary reported : {Count}", subsidiaries.Count().ToString());
```
[the remark this answers is not in the document]
Comment (abidkayanidefra) [reply src/EPR.SubsidiaryBulkUpload.Application/Services/BulkSubsidiaryProcessor.cs]: done

**Pull request merged**